test/foundry/utils/version-control/RoninValidatorSetTimedMigrator.sol

```diff
@@ -0,0 +1,175 @@
+
+    vm.prank(admin, admin);
+
+    proxyStorage = payable(address(new TransparentUpgradeableProxyV2(logicV1, admin, "")));
+    versionController = payable(address(new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, logicV2)));
+    bobContract = new MockActor(proxyStorage);
+  }
+
+  function testManualUpgradeToVersionControl() public {
+    _testManualUpgrade(versionController);
+  }
+
+  function _testManualUpgrade(address implementation) private {
+    vm.expectEmit(true, false, false, false);
+    emit Upgraded(implementation);
+    vm.startPrank(admin, admin);
+    TransparentUpgradeableProxyV2(proxyStorage).upgradeTo(implementation);
+    assertEq(implementation, TransparentUpgradeableProxyV2(proxyStorage).implementation());
+    vm.stopPrank();
+  }
+
+  function testFailDupplicateInput0() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, proxyStorage, proxyStorage);
+  }
+
+  function testFailDupplicateInput1() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, logicV1);
+  }
+
+  function testFailDupplicateInput2() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(logicV1, proxyStorage, logicV1);
+  }
+
+  function testFailNonContractInput0() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, alice));
+    new RoninValidatorSetTimedMigrator(alice, logicV1, logicV2);
+  }
+
+  function testFailNonContractInput1() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, alice));
+    new RoninValidatorSetTimedMigrator(proxyStorage, alice, logicV2);
+  }
+
+  function testFailNonContractInput2() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, alice));
+    new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, alice);
+  }
+
+  function testFailNullInput0() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, address(0)));
+    new RoninValidatorSetTimedMigrator(address(0), logicV1, logicV2);
+  }
+
+  function testFailNullInput1() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, address(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, address(0), logicV2);
+  }
+
+  function testFailNullInput2() public {
+    vm.expectRevert(abi.encodePacked(ErrZeroCodeContract.selector, address(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, address(0));
+  }
+
+  function testGetLogicV1AfterUpgradeToVersionControl() public {
+    testManualUpgradeToVersionControl();
+    string memory name = ILogicValidatorSet(proxyStorage).version();
+    assertEq(name, "V1");
+  }
+
+  function testGetLogicV2AfterUpgradeToVersionControl() public {
+    testGetLogicV1AfterUpgradeToVersionControl();
+    vm.roll(100);
```

Comment:
Why 100? If you want to use snapshot, use `snapshot()` and `revertTo(snapshotId)` instead

Reply:
forward to block.number >= 100 to test migration functionality

Reply:
Then it should be:
```solidity
function _mineBlocks(uint256 n) internal {
  vm.roll(block.number + n);
}
```

```diff
-    vm.roll(100);
+    _mineBlocks(100);
```

Reply:
I dont think we need to wrap it in `_mineBlocks`. `vm.roll` is common pattern when using foundry test. Just treat it as `vm.prank`, `vm.xxx`.